### ISSUE TITLE
gstreamer v3.5.0 - signal_connect and minitest side effects 

### ISSUE THREAD

**Issue by mcclumpherty** [ruby-gnome/ruby-gnome#1445]
This has to be one of the weirdest ones I've ever seen.

TL:DR - signal handlers within minitests don't work depending on the order of the `require` statements within the test script

### Environment
My environment is Linux Mint 19.3, Ruby 3.0.1 and using the 3.5.0 branch of ruby-gnome.

### Description
I was interested in using signal handlers and decided to use the sample [typefind.rb](https://github.com/ruby-gnome/ruby-gnome/blob/master/gstreamer/sample/typefind.rb) code as a basis for this.

The plan was to set up a signal handler for the `typefind` element's `have-type` signal to capture the caps for a given file.  To test this, I wrote a minitest that used ffmpeg to generate a bunch of test video files and run the modified sample against these files.

### Code
The code is structured as follows:
```
typefinder/
  |- Gemfile
  |- medial-tools/
       |- typefinder.rb
  |- test/
      |- typefinder_test.rb
```

### Gemfile
```ruby
source "https://rubygems.org"

gem 'minitest'

git "https://github.com/ruby-gnome/ruby-gnome.git" do
  gem "gstreamer"
end

```

### typefind.rb
To demonstrate this, I took the sample code and converted it into a class so that it could be called from a minitest.  I have littered it with `puts` statements to demonstrate what is going on.

```ruby
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
#
# This sample code is a port of
# gstreamer/tests/examples/typefind/typefind.c. It is licensed
# under the terms of the GNU Library General Public License, version
# 2 or (at your option) later.
#
# Here is the original copyright header:
#   Copyright (C) <2005> Stefan Kost
#   Copyright (C) <2006> Tim-Philipp Müller
#
# Copyright (C) 2013  Ruby-GNOME2 Project Team
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

require "gstreamer"

puts "gstremer gem version #{Gem.loaded_specs['gstreamer'].version}"

class Typefinder

  def self.event_loop(pipe, typefind)
    running = true
    bus = pipe.bus

    puts "#{File.basename(__FILE__ )}:#{__LINE__} Running event loop"

    while running do
      message = bus.poll(Gst::MessageType::ANY, Gst::CLOCK_TIME_NONE)
      raise "message nil" if message.nil?

      signal_handler_pending = typefind.signal_has_handler_pending?('have-type')
      puts "#{File.basename(__FILE__ )}:#{__LINE__} signal_has_handler_pending? #{signal_handler_pending}"

      case message.type
      when Gst::MessageType::EOS
        puts "#{File.basename(__FILE__ )}:#{__LINE__} #{message.src.name} -> EOS"
        running = false
      when Gst::MessageType::STATE_CHANGED
        old, new, pend = message.parse_state_changed
        puts "#{File.basename(__FILE__ )}:#{__LINE__} #{message.src.name} #{message.type.nick} -> new state=#{new.nick}"
      when Gst::MessageType::WARNING
        warning, debug = message.parse_warning
        puts "#{File.basename(__FILE__ )}:#{__LINE__} #{message.src.name} #{message.type.nick} -> warning=#{warning.message}"
      when Gst::MessageType::ERROR
        error, debug = message.parse_error
        puts "#{File.basename(__FILE__ )}:#{__LINE__} #{message.src.name} #{message.type.nick} -> error=#{error.message}"
        running = false
      else
        puts "#{File.basename(__FILE__ )}:#{__LINE__} #{message.src.name} #{message.type.nick}"
      end
    end
  end

  def self.typefind(filename)

    caps_str = nil

    raise "File not found: #{filename}" unless File.exist?(filename)

    # create a new pipeline to hold the elements
    pipeline = Gst::Pipeline.new("pipeline")
    raise "'pipeline' gstreamer plugin missing" if pipeline.nil?

    # create a file reader
    filesrc = Gst::ElementFactory.make("filesrc", "file_source")
    raise "'filesrc' gstreamer plugin missing" if filesrc.nil?
    filesrc.location = filename

    typefind = Gst::ElementFactory.make("typefind", "typefind")
    raise "'typefind' gstreamer plugin missing" if typefind.nil?

    sink = Gst::ElementFactory.make("fakesink", "sink")
    raise "'fakesink' gstreamer plugin missing" if sink.nil?

    # add objects to the main pipeline
    pipeline << filesrc << typefind << sink

    handler_id = typefind.signal_connect("have-type") do |element, probability, caps|
      caps_str = caps.to_s
      puts "#{File.basename(__FILE__ )}:#{__LINE__} Signal handler called for #{element.name} - got caps [#{caps_str}]"
      puts "#{File.basename(__FILE__ )}:#{__LINE__} signal_has_handler_pending? #{typefind.signal_has_handler_pending?('have-type')}"
    end

    puts "#{File.basename(__FILE__ )}:#{__LINE__} Added signal handler (handler_id=#{handler_id})"

    puts "#{File.basename(__FILE__ )}:#{__LINE__} signal_has_handler_pending? #{typefind.signal_has_handler_pending?('have-type')}"

    filesrc >> typefind >> sink

    # start playing
    puts "#{File.basename(__FILE__ )}:#{__LINE__} Playing pipeline"
    pipeline.play

    # Run event loop listening for bus messages until EOS or ERROR
    event_loop(pipeline, typefind)

    # stop the bin
    pipeline.stop

    puts "#{File.basename(__FILE__ )}:#{__LINE__} signal_has_handler_pending? #{typefind.signal_has_handler_pending?('have-type')}"

    caps_str
  end
end
```

### typefinder_test.rb
And now here is the minitest:
```ruby
# THIS WORKS
require './media-tools/typefinder'
require 'minitest/autorun'

# THIS DOESN'T WORK
# require 'minitest/autorun'
# require './media-tools/typefinder'

class TypefinderTest < Minitest::Test

  @@test_videos = nil

  def setup
    if @@test_videos.nil?
      @@test_videos = []
      ['avi', 'flv', 'mkv', 'mov', 'mp4', 'ts', 'wmv'].each do |ext|
        input_video = "-f lavfi -i testsrc=d=2:size=704x576:r=25 -c:v flv"
        input_audio = "-f lavfi -i sine=f=440:b=4 -shortest -c:a aac"
        output_file = "/tmp/gstreamer-test-video.#{ext}"
        cmd = "ffmpeg -v error -y -loglevel quiet -nostats -hide_banner #{input_video} #{input_audio} #{output_file}"

        unless File.exist?(output_file)
          puts "#{File.basename(__FILE__ )}:#{__LINE__} Creating test video #{output_file}"
          system cmd
        end

        @@test_videos << output_file
      end
    end
  end

  def teardown
    # Do nothing
  end

  def test_typefind
    caps = Typefinder.typefind(@@test_videos[0])
    assert !caps.nil?
  end
end
```
### Successful Output
Here is the output from running the test where the `require` statements are in this order:
```ruby
require './media-tools/typefinder'
require 'minitest/autorun'
```

```
gstremer gem version 3.5.0
typefinder.rb:98 Added signal handler (handler_id=1)
typefinder.rb:100 signal_has_handler_pending? true
typefinder.rb:105 Playing pipeline
typefinder.rb:39 Running event loop
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 sink state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 typefind state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 file_source state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 pipeline state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:63 sink stream-status
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 typefind state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 file_source state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:63 sink stream-status
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:63 pipeline stream-start
typefinder.rb:94 Signal handler called for typefind - got caps [video/x-msvideo]
typefinder.rb:95 signal_has_handler_pending? true
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 sink state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 pipeline state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:63 pipeline async-done
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:63 pipeline new-clock
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 sink state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 typefind state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 file_source state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 pipeline state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:50 pipeline -> EOS
typefinder.rb:114 signal_has_handler_pending? true
```

Note that throughout the run, the signal handler remains in place and the signal handler for the typefind 'have-type' signal is called.

### Side-Effects Output
Ok, here is the output for when it is run where the `require` statements are in this order:
```
require 'minitest/autorun'
require 'media-tools/typefinder'
```

```
gstremer gem version 3.5.0
typefinder.rb:98 Added signal handler (handler_id=1)
typefinder.rb:100 signal_has_handler_pending? true
typefinder.rb:105 Playing pipeline
typefinder.rb:39 Running event loop
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 sink state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 typefind state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 file_source state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 pipeline state-changed -> new state=ready
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:63 sink stream-status
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 typefind state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? true
typefinder.rb:54 file_source state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:63 sink stream-status
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:63 pipeline stream-start
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:54 sink state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:54 pipeline state-changed -> new state=paused
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:63 pipeline async-done
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:63 pipeline new-clock
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:54 sink state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:54 typefind state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:54 file_source state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:54 pipeline state-changed -> new state=playing
typefinder.rb:46 signal_has_handler_pending? false
typefinder.rb:50 pipeline -> EOS
typefinder.rb:114 signal_has_handler_pending? false

Minitest::Assertion: Expected false to be truthy.
```

You'll see that once the paused state is reached, `typefind.signal_has_handler_pending?('have-type')` starts to return `false` and the signal handler never gets called.

### Conclusion
This is just bonkers.   

Why would requiring `minitools/autorun` before the class under test cause the signal handler to disappear when the paused state is reached?

Any ideas why this might be the case?

Archie

**Comment by mcclumpherty** [comment]
That's great to know.  Many thanks for the explanation !
